### ISSUE TITLE
[BUG] SD card is not reading

### ISSUE THREAD

**Issue by FreeZuu69** [MarlinFirmware/Marlin#24712]
### Did you test the latest `bugfix-2.1.x` code?

Yes, and the problem still exists.

### Bug Description

SD card is not reading.

my system

Mks Gen L v2.1
Mks H43 Display
Mks H43 Adapter
Mks PWC v3.0

![112](https://user-images.githubusercontent.com/108133167/187859363-6ee7afd3-83a0-4785-902e-caa506d44b05.png)

### Bug Timeline

a new error, it does not read the card in any way.

### Expected behavior

As soon as the printer is turned on, the contents of the sd card should appear, but the files are not visible. There is no such problem in the previous version. Marlin-V2.X-MKS-H43-main

### Version of Marlin Firmware

Marlin-2.1.1

### Printer model

Corexy

### Electronics

Mks  gen L 2.1 , Mks Pwc, Mks H43 LCD, bltouch, h43 a adapter

### Bed Leveling

ABL Bilinear mesh

### Your Slicer

Prusa Slicer

### Host Software

SD Card (headless)

### Don't forget to include

- [X] A ZIP file containing your `Configuration.h` and `Configuration_adv.h`.

### Additional information & file uploads

[Marlin.zip](https://github.com/MarlinFirmware/Marlin/files/9467841/Marlin.zip)

**Comment by borland1** [comment]
Disabled by default...

Configuration.h
 ```
* SD CARD
 *
 * SD Card support is disabled by default. If your controller has an SD slot,
 * you must uncomment the following option or it won't work.
 */
//#define SDSUPPORT

/**
 * SD CARD: ENABLE CRC
 *
 * Use CRC checks and retries on the SD communication.
 */
//#define SD_CHECK_AND_RETRY
```

**Comment by FreeZuu69** [comment]
I tried both of them differently. nothing has changed, the problem persists.

**Comment by ellensp** [comment]
Provided Config does not compile
```
Compiling .pio/build/mega2560/src/src/core/serial.cpp.o
In file included from Marlin/src/HAL/AVR/../../inc/MarlinConfig.h:49:0,
                 from Marlin/src/HAL/AVR/HAL.cpp:24:
Marlin/src/HAL/AVR/../../inc/SanityCheck.h:3186:4: error: #error "TMC2208 or TMC2209 on Z2 requires Z2_HARDWARE_SERIAL or Z2_SERIAL_(RX|TX)_PIN."
```

Updated to current bugfix configs and added Z2 pins

[Configuration-2.0.x.zip](https://github.com/MarlinFirmware/Marlin/files/9662797/Configuration-2.0.x.zip)

**Comment by ellensp** [comment]
First issue I see 
```
PIN:  12   Port: B6        E1_CS_PIN                              protected
.                          PS_ON_PIN                              protected
.                          Z2_SERIAL_RX_PIN                       protected
.                          E1_SERIAL_RX_PIN                       protected

```
So there is a conflict with  PS_ON_PIN and Z2_SERIAL_RX_PIN that needs resolved.
No idea where it should be so putting it on pin 4 for now.
  #define PS_ON_PIN                            4

**Comment by ellensp** [comment]
no sd detect pin either
  #define SD_DETECT_PIN                       49

**Comment by ellensp** [comment]
Odd behavior during testing, with the MKS H43 Adapter-A connected  and the SD_DETECT_PIN defined

If you disable DGUS_LCD_UI_MKS and enable REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
The sd card on the MKS H43 Adapter-A  works as expected

But when you enable DGUS_LCD_UI_MKS and disable REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
The sd card on the MKS H43 Adapter-A  fails to work...   

If your connected via serial and insert a sdcard the controller seem to do a partial reboot and then flashes the led (pin 13) forever. 
Even even resetting the micro with the reset button does not help.
You have to power down after this.

**Comment by ellensp** [comment]
It seem to be a memory issue...  to many things enabled on old mega2560 with only 4k 
If you for example change TMC2209 to TMC2209_STANDALONE, without those 5 serial buffers it frees up enough ram to work as expected.

**Comment by jmger** [comment]
YES, I have the same issue. I have the Creality Ender 3 Max with BTT Octopus Pro F429 board. I enabled SDSUPPORT and CRC check. The SDCARD was working in 2.0.x. My BTT BOARD uploaded and installed the 2.1.2 Firmware file just fine. Once or twice I get the message Initialization fail. For the SDCARD.

**Comment by jmger** [comment]
HI for at least my issue where SDCARD was not detected/working in Marlin 2.1.2. I had to add the line #define SDCARD_CONNECTION ONBOARD to the Configuration.adv file. Apparently, This line was left out. Now everything is working.

**Comment by inspectionsbybob** [comment]
Forgive me, but I am having similar problems with the MKS_Robin_Pro

When I look at the Config_Adv.H file I can find nothing mentioning the "# define SDcard_CONNECTION_ONBOARD" and normally there is a "If SDcard_CONNECTION_OMBOARD " followed by a whole lot of defines and other stuff. Where did you find the code that uses the flag you say is missing?

**Comment by agolinko** [comment]
Not working for me too. 
After boot, contents of SDCARD is visible. But when select something to print, after confirm, it does nothing. When trying to open SD card again it shows empty card.

Hardware: Kingroon K3PS/Robin nano board,  BLTouch
Marlin 2.1.2.1

**Comment by bhargavbe** [comment]
I am also facing the same issue now, suddenly !!!, was working fine before,  in 2.1.2.1 branch, It had another issue of stopping print intermediately without any reason.
Now print doesn't even start. 
For additional information.
1. I have tried to print from SD card from menu. 
2. I have tried to print from SD card from prontoface.
3. I have tried to print from prontoface as a host and selected file from my PC.
 In all above scenario "Not even first line execute". So it should not be the issue specific in SD Card but it's the issue related to print.
 
 All commands sent from serial port directly works fine,

Hardware: Ramps 1.4, Mega 2560
[Config.zip](https://github.com/MarlinFirmware/Marlin/files/12386174/Config.zip)

**Comment by mechano** [comment]
> Not working for me too. After boot, contents of SDCARD is visible. But when select something to print, after confirm, it does nothing. When trying to open SD card again it shows empty card.
> 
> Hardware: Kingroon K3PS/Robin nano board, BLTouch Marlin 2.1.2.1

Same beaviour, everything perfect and fully working but the SD reader. Files appears I select one and it doesn't print. When I go back to media the files disappear. If I extract and insert again the SD card they appeat again but still no printing. Tryed both 2.1.2.1 and 2.1.2. 

I saw around 2.1.2 firmwares fully working. So I think it's only a misconfiguration.

**Comment by ElMedic0** [comment]
I had the same issue, fixed by uncommenting those if you are using the lcd sdcard slot,
//#define SDCARD_CONNECTION LCD
//#define NO_SD_DETECT 
And also left the power loss commented (disabled).
//#define POWER_LOSS_RECOVERY
